user/overriding-packages.rst

```diff
@@ -0,0 +1,280 @@
+It is accomplished by sourcing an existing workspace, then building a different version of that package again in a new overlay workspace.
+
+Overriding a package is not always possible.
+This page offers tips for avoiding common issues.
+
+.. contents:: Table of Contents
+    :depth: 3
+
+
+How to avoid common issues
+--------------------------
+
+These are good practices to avoid common issues.
+If the advice is too restrictive then see the known issue descriptions for more complex advice.
+
+Use isolated workspaces
+***********************
+
+If you are building a workspace and suspect you may overriding a package from it in the future, then use an isolated workspace.
+This avoids include directory search order issues.
+See :doc:`this documentation <isolated-vs-merged-workspaces>` if you're unsure what an isolated workspace is.
+
+Override every package that depends on the one you want to override
+*******************************************************************
+
+A **leaf package** is one that has no other packages that depend on it.
+Overriding a non-leaf package can be problematic.
+Packages in the underlay were built against the underlay version of the package, but they will be expected to run with the overlay version.
+If their build process stores some information about the non-leaf package, such as an expected ABI, then undefined behavior can happen at runtime.
+
+Problems caused by packages remembering information at build time can be avoided by overriding every package that directly or indirectly depends on the one you actually want to override.
+The group of overridden packages must span all underlays.
+
+Say there are 3 workspaces, **A**, **B** and **C**, where **C** overlays **B** and **B** overlays **A**.
+**A** contains packages ``foo`` and ``baz`` where ``baz`` depends on ``foo``.
+**B** contains packages ``ping`` that depends on ``foo`` and  ``pong`` that depends on ``baz``.
+**C** is the workspace being built.
+If you want to override ``foo`` then you should also override ``baz``, ``ping``, and ``pong``.
+
+Make sure overridden Python packages do not change entry point specifications
+*****************************************************************************
+
+Python packages may have `entry point specifications <https://packaging.python.org/en/latest/specifications/entry-points/>`_ in a ``setup.py`` or ``setup.cfg`` file.
+Make sure the package in the overlay has identical specifications to the version in the underlay, or only adds new ones.
+If any specification has been changed or removed then it may not be possible to override this package.
+
+How to make it easier for your users to override
+------------------------------------------------
+
+This section has advice for package authors about how to make easier for your users to use your package and override it or other pacakges.
+
+Install headers to a unique include directory
+*********************************************
```

Comment:
This section is good. After talking with you a few weeks back I wrote up a dive into octomap_msgs on a project. Here's that excerpted in case you like any of the talk-through info to add to this section.

> The exported include dir includes the package name, but in order for the header to also  include the package name that package name directory has to be doubly nested.
> To use `include/$PACKAGENAME` as the include directory and use `#include <$PACKAGENAME/header.h>`  in your files, the directory path has to be
`include/$PACKAGENAME/$PACKAGENAME/header.h`

> For ROS packages we want `$PACKAGENAME` in the include path so that include paths in merged workspaces don't overlap, and are thus override-able in overlay workspaces.
In order to keep the package origins of our headers straight in source files, we also want the package name in our `#include` lines.
So we can't get rid of either of them without creating confusion and breakage somewhere.
Even though for isolated workspaces `install/$PACKAGENAME/include/$PACKAGENAME/$PACKAGENAME/header.h` gives me real `User user = new User()` Java flashbacks.

> Since `octomap_msgs_INCLUDE_DIRS` / `octomap_msgs_TARGETS` come from ament_package() which is generating them on behalf of the interfaces, I don't know what the "correct" way for an interface package to install additional non-generated interface headers is. But I think the proposed patch, which installs the custom header alongside the interface headers in `$prefix/include/$PACKAGENAME` rather than in `$prefix/include` is an improvement and is more correct than exporting an additional `INCLUDE_DIRS` / `TARGETS` element for `$prefix/include`